src/pgen/cluster/agn_feedback.cpp

```diff
@@ -0,0 +1,230 @@
+//========================================================================================
+//! \file agn_feedback.cpp
+//  \brief  Class for injecting AGN feedback via thermal dump, kinetic jet, and magnetic tower
+
+// Parthenon headers
+#include <coordinates/uniform_cartesian.hpp>
+#include <globals.hpp>
+#include <interface/state_descriptor.hpp>
+#include <mesh/domain.hpp>
+#include <parameter_input.hpp>
+#include <parthenon/package.hpp>
+
+// Athena headers
+#include "../../eos/adiabatic_glmmhd.hpp"
+#include "../../eos/adiabatic_hydro.hpp"
+#include "../../main.hpp"
+#include "../../units.hpp"
+#include "agn_feedback.hpp"
+#include "agn_triggering.hpp"
+#include "cluster_utils.hpp"
+#include "magnetic_tower.hpp"
+
+namespace cluster {
+using namespace parthenon;
+
+AGNFeedback::AGNFeedback(parthenon::ParameterInput *pin,
+                         parthenon::StateDescriptor *hydro_pkg)
+    : fixed_power_(pin->GetOrAddReal("problem/cluster/agn_feedback", "fixed_power", 0.0)),
+      efficiency_(pin->GetOrAddReal("problem/cluster/agn_feedback", "efficiency", 1e-3)),
+      thermal_fraction_(
+          pin->GetOrAddReal("problem/cluster/agn_feedback", "thermal_fraction", 0.0)),
+      kinetic_fraction_(
+          pin->GetOrAddReal("problem/cluster/agn_feedback", "kinetic_fraction", 0.0)),
+      magnetic_fraction_(
+          pin->GetOrAddReal("problem/cluster/agn_feedback", "magnetic_fraction", 0.0)),
+      thermal_radius_(
+          pin->GetOrAddReal("problem/cluster/agn_feedback", "thermal_radius", 0.01)),
+      kinetic_jet_radius_(
+          pin->GetOrAddReal("problem/cluster/agn_feedback", "kinetic_jet_radius", 0.01)),
+      kinetic_jet_height_(
+          pin->GetOrAddReal("problem/cluster/agn_feedback", "kinetic_jet_height", 0.02)),
+      disabled_(pin->GetOrAddBoolean("problem/cluster/agn_feedback", "disabled", false)) {
+
+  //Normalize the thermal, kinetic, and magnetic fractions to sum to 1.0
+  const Real total_frac = thermal_fraction_ + kinetic_fraction_ + magnetic_fraction_;
+  thermal_fraction_ = thermal_fraction_/total_frac;
+  kinetic_fraction_ = kinetic_fraction_/total_frac;
+  magnetic_fraction_ = magnetic_fraction_/total_frac;
+
+  // Add user history output variable for AGN power
+  auto hst_vars = hydro_pkg->Param<parthenon::HstVar_list>(parthenon::hist_param_key);
+  if (!disabled_) {
+    // HACK (forrestglines): The operations should be a
+    // parthenon::UserHistoryOperation::no_reduce, which is as of writing
+    // unimplemented
+    hst_vars.emplace_back(parthenon::HistoryOutputVar(
+        parthenon::UserHistoryOperation::max,
+        [this](MeshData<Real> *md) {
+          auto pmb = md->GetBlockData(0)->GetBlockPointer();
+          auto hydro_pkg = pmb->packages.Get("Hydro");
+          return this->GetFeedbackPower(hydro_pkg.get());
+        },
+        "agn_feedback_power"));
+  }
+  hydro_pkg->UpdateParam(parthenon::hist_param_key, hst_vars);
```

Comment:
Yes, we do want the AGN feedback power in the history output. It's not necessary for simulation correctness but it would be essentially necessary for analysis.

There was discussion of implementing a `no_reduce` history operation that wouldn't do any reduction/barrier for this variable.

I think that said performance fix, especially for the other variables, would be great. Theoretically these could all be nonblocking reductions with one barrier after all the history reductions?